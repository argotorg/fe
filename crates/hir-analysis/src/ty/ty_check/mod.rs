mod env;
mod expr;
mod pat;
mod stmt;

use env::TyCheckEnv;
use hir::{
    hir_def::{
        Body, Enum, ExprId, Func, LitKind, PatId, PathId, TypeId as HirTyId,
        VariantKind as HirVariantKind,
    },
    span::DynLazySpan,
};
use rustc_hash::FxHashMap;

use super::{
    diagnostics::{BodyDiag, FuncBodyDiagAccumulator},
    ty_def::{AdtDef, AdtRef, AdtRefId, InvalidCause, Kind, TyId, TyVarUniverse},
    ty_lower::{lower_adt, lower_hir_ty},
    unify::{UnificationError, UnificationTable},
};
use crate::HirAnalysisDb;

#[salsa::tracked(return_ref)]
pub fn check_func_body(db: &dyn HirAnalysisDb, func: Func) -> TypedBody {
    let Ok(mut checker) = TyChecker::new_with_func(db, func) else {
        return TypedBody::empty();
    };

    checker.run();
    checker.finish()
}

pub struct TyChecker<'db> {
    db: &'db dyn HirAnalysisDb,
    env: TyCheckEnv<'db>,
    table: UnificationTable<'db>,
    expected: TyId,
}

impl<'db> TyChecker<'db> {
    fn new_with_func(db: &'db dyn HirAnalysisDb, func: Func) -> Result<Self, ()> {
        let env = TyCheckEnv::new_with_func(db, func)?;
        let expected_ty = match func.ret_ty(db.as_hir_db()) {
            Some(hir_ty) => {
                let ty = lower_hir_ty(db, hir_ty, func.scope());
                if ty.is_star_kind(db) {
                    ty
                } else {
                    TyId::invalid(db, InvalidCause::Other)
                }
            }
            None => TyId::unit(db),
        };

        Ok(Self::new(db, env, expected_ty))
    }

    fn run(&mut self) {
        let root_expr = self.env.body().expr(self.db.as_hir_db());
        self.check_expr(root_expr, self.expected);
    }

    fn finish(mut self) -> TypedBody {
        // TODO: check for untyped expressions and patterns.
        self.env.finish(&mut self.table)
    }

    fn new(db: &'db dyn HirAnalysisDb, env: TyCheckEnv<'db>, expected: TyId) -> Self {
        let table = UnificationTable::new(db);
        Self {
            db,
            env,
            table,
            expected,
        }
    }

    fn body(&self) -> Body {
        self.env.body()
    }

    fn lit_ty(&mut self, lit: &LitKind) -> TyId {
        match lit {
            LitKind::Bool(_) => TyId::bool(self.db),
            LitKind::Int(_) => self.table.new_var(TyVarUniverse::Integral, &Kind::Star),
            LitKind::String(s) => {
                let len_bytes = s.len_bytes(self.db.as_hir_db());
                self.table
                    .new_var(TyVarUniverse::String(len_bytes), &Kind::Star)
            }
        }
    }

    fn lower_ty(&self, hir_ty: HirTyId, span: DynLazySpan) -> TyId {
        let ty = lower_hir_ty(self.db, hir_ty, self.env.scope());
        if let Some(diag) = ty.emit_diag(self.db, span) {
            FuncBodyDiagAccumulator::push(self.db, diag.into());
        }

        ty
    }

    /// Returns the fresh type variable for pattern and expr type checking. The
    /// kind of the type variable is `*`, and the universe is `General`.
    fn fresh_ty(&mut self) -> TyId {
        self.table.new_var(TyVarUniverse::General, &Kind::Star)
    }

    fn fresh_tys_n(&mut self, n: usize) -> Vec<TyId> {
        (0..n).map(|_| self.fresh_ty()).collect()
    }

    fn unify_ty<T>(&mut self, t: T, actual: TyId, expected: TyId) -> TyId
    where
        T: Into<Typeable>,
    {
        let t = t.into();

        let actual = match self.table.unify(expected, actual) {
            Ok(()) => actual,

            Err(UnificationError::TypeMismatch) => {
                let actual = actual.apply_subst(self.db, &mut self.table);
                let expected = expected.apply_subst(self.db, &mut self.table);
                FuncBodyDiagAccumulator::push(
                    self.db,
                    BodyDiag::type_mismatch(
                        self.db,
                        t.lazy_span(self.env.body()),
                        expected,
                        actual,
                    )
                    .into(),
                );
                TyId::invalid(self.db, InvalidCause::Other)
            }

            Err(UnificationError::OccursCheckFailed) => {
                FuncBodyDiagAccumulator::push(
                    self.db,
                    BodyDiag::InfiniteOccurrence(t.lazy_span(self.env.body())).into(),
                );

                TyId::invalid(self.db, InvalidCause::Other)
            }
        };

        match t {
            Typeable::Expr(expr) => self.env.type_expr(expr, actual),
            Typeable::Pat(pat) => self.env.type_pat(pat, actual),
        }

        actual
    }
}

#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TypedBody {
    body: Option<Body>,
    pat_ty: FxHashMap<PatId, TyId>,
    expr_ty: FxHashMap<ExprId, TyId>,
}

impl TypedBody {
    pub fn expr_ty(&self, db: &dyn HirAnalysisDb, expr: ExprId) -> TyId {
        self.expr_ty
            .get(&expr)
            .copied()
            .unwrap_or_else(|| TyId::invalid(db, InvalidCause::Other))
    }
    pub fn pat_ty(&self, db: &dyn HirAnalysisDb, pat: PatId) -> TyId {
        self.pat_ty
            .get(&pat)
            .copied()
            .unwrap_or_else(|| TyId::invalid(db, InvalidCause::Other))
    }

    fn empty() -> Self {
        Self {
            body: None,
            pat_ty: FxHashMap::default(),
            expr_ty: FxHashMap::default(),
        }
    }
}

#[derive(Clone, Copy, PartialEq, Eq, derive_more::From)]
enum Typeable {
    Expr(ExprId),
    Pat(PatId),
}

impl Typeable {
    fn lazy_span(self, body: Body) -> DynLazySpan {
        match self {
            Self::Expr(expr) => expr.lazy_span(body).into(),
            Self::Pat(pat) => pat.lazy_span(body).into(),
        }
    }
}

#[derive(Clone, Debug, Hash, PartialEq, Eq)]
pub(crate) enum ResolvedPathData {
    Adt(AdtDef, PathId),
    Variant(Enum, usize, PathId),
}

impl ResolvedPathData {
    pub(crate) fn adt_ref(&self, db: &dyn HirAnalysisDb) -> AdtRef {
        match self {
            Self::Adt(adt, _) => adt.adt_ref(db).data(db),
            Self::Variant(enum_, _, _) => AdtRef::Enum(*enum_),
        }
    }

    pub(crate) fn data_kind(&self, db: &dyn HirAnalysisDb) -> &'static str {
        match self {
            Self::Adt(adt, _) => adt.adt_ref(db).kind_name(db),
            Self::Variant(enum_, idx, _) => {
                let hir_db = db.as_hir_db();
                match enum_.variants(hir_db).data(hir_db)[*idx].kind {
                    hir::hir_def::VariantKind::Unit => "unit variant",
                    HirVariantKind::Tuple(_) => "tuple variant",
                    HirVariantKind::Record(_) => "record variant",
                }
            }
        }
    }

    pub(crate) fn initializer_hint(&self, db: &dyn HirAnalysisDb) -> Option<String> {
        let hir_db = db.as_hir_db();

        let expected_sub_pat = match self {
            Self::Adt(_, _) => {
                let AdtRef::Struct(s) = self.adt_ref(db) else {
                    return None;
                };

                s.format_initializer_args(hir_db)
            }

            Self::Variant(enum_, idx, _) => {
                enum_.variants(hir_db).data(hir_db)[*idx].format_initializer_args(hir_db)
            }
        };

        let path = self.path().pretty_print(hir_db);
        Some(format!("{}{}", path, expected_sub_pat))
    }

    fn path(&self) -> PathId {
        match self {
            Self::Adt(_, path) => *path,
            Self::Variant(_, _, path) => *path,
        }
    }

    fn ty(&self, db: &dyn HirAnalysisDb, table: &mut UnificationTable) -> TyId {
        let adt = match self {
            Self::Adt(adt, _) => *adt,

            Self::Variant(enum_, ..) => lower_adt(db, AdtRefId::from_enum(db, *enum_)),
        };

        let adt_ty = TyId::adt(db, adt);
        adt.params(db).iter().fold(adt_ty, |ty, param| {
            let param_ty = table.new_var_from_param(*param);
            TyId::app(db, ty, param_ty)
        })
    }

    fn is_unit_variant(&self, db: &dyn HirAnalysisDb) -> bool {
        match self {
            Self::Adt(_, _) => false,
            Self::Variant(enum_, idx, _) => {
                let hir_db = db.as_hir_db();
                matches!(
                    enum_.variants(hir_db).data(hir_db)[*idx].kind,
                    hir::hir_def::VariantKind::Unit
                )
            }
        }
    }

    fn is_tuple_variant(&self, db: &dyn HirAnalysisDb) -> bool {
        match self {
            Self::Adt(_, _) => false,
            Self::Variant(enum_, idx, _) => {
                let hir_db = db.as_hir_db();
                matches!(
                    enum_.variants(hir_db).data(hir_db)[*idx].kind,
                    HirVariantKind::Tuple(_)
                )
            }
        }
    }

    fn field_tys(&self, db: &dyn HirAnalysisDb) -> Vec<TyId> {
        let (adt, idx) = match self {
            Self::Adt(adt, _) => {
                if matches!(
                    adt.adt_ref(db).data(db),
                    AdtRef::Struct(_) | AdtRef::Contract(_)
                ) {
                    (*adt, 0)
                } else {
                    return vec![];
                }
            }
            Self::Variant(enum_, idx, _) => {
                let adt = lower_adt(db, AdtRefId::from_enum(db, *enum_));
                (adt, *idx)
            }
        };

        adt.fields(db)[idx].iter_types(db).collect()
    }
}
