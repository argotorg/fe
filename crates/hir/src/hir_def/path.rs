use super::{kw, IdentId};
use crate::{hir_def::Partial, HirDb};

#[salsa::interned]
pub struct PathId {
    #[return_ref]
    pub segments: Vec<Partial<IdentId>>,
}

impl PathId {
    pub fn last_segment(self, db: &dyn HirDb) -> Partial<IdentId> {
        self.segments(db).last().copied().unwrap_or_default()
    }

    pub fn len(self, db: &dyn HirDb) -> usize {
        self.segments(db).len()
    }

    pub fn is_ident(self, db: &dyn HirDb) -> bool {
        self.len(db) == 1
    }

    pub fn self_ty(db: &dyn HirDb) -> Self {
        let self_ty = Partial::Present(kw::SELF_TY);
        Self::new(db, vec![self_ty])
    }

    pub fn from_ident(db: &dyn HirDb, ident: IdentId) -> Self {
        Self::new(db, vec![Partial::Present(ident)])
    }
}
