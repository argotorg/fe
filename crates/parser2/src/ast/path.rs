use rowan::ast::{support, AstNode};

use super::{ast_node, AstChildren};
use crate::{syntax_node::SyntaxToken, SyntaxKind as SK};

ast_node! {
    /// A path.
    /// `foo::bar::baz`
    pub struct Path,
    SK::Path,
    IntoIterator<Item=PathSegment>,
}
impl Path {
    /// Returns the segments of the path.
    pub fn segments(&self) -> AstChildren<PathSegment> {
        support::children(self.syntax())
    }
}

ast_node! {
    /// A path segment.
    pub struct PathSegment,
    SK::PathSegment
}
impl PathSegment {
    /// Returns the identifier of the segment.
    pub fn ident(&self) -> Option<SyntaxToken> {
        support::token(self.syntax(), SK::Ident)
    }

    /// Returns `true` if the segment  is a `self` keyword.
    pub fn is_self(&self) -> bool {
        support::token(self.syntax(), SK::SelfKw).is_some()
    }

    /// Returns `true` if the segment is a `Self` keyword.
    pub fn is_self_ty(&self) -> bool {
        support::token(self.syntax(), SK::SelfTypeKw).is_some()
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::{
        lexer::Lexer,
        parser::{path::PathScope, Parser},
    };

    use wasm_bindgen_test::wasm_bindgen_test;

    fn parse_path(source: &str) -> Path {
        let lexer = Lexer::new(source);
        let mut parser = Parser::new(lexer);
        parser.parse(PathScope::default(), None);
        Path::cast(parser.finish().0).unwrap()
    }

    #[test]
    #[wasm_bindgen_test]
    fn path_ast() {
        let source = r#"self::Foo"#;
        let path = parse_path(source);
        let mut segments = path.segments();

        assert!(segments.next().unwrap().is_self());
        assert_eq!(segments.next().unwrap().ident().unwrap().text(), "Foo");
        assert!(segments.next().is_none());
    }

    #[test]
    #[wasm_bindgen_test]
    fn path_ast2() {
        let source = r#"Self::Dep"#;
        let path = parse_path(source);
        let mut segments = path.segments();

        assert!(segments.next().unwrap().is_self_ty());
        assert_eq!(segments.next().unwrap().ident().unwrap().text(), "Dep");
        assert!(segments.next().is_none());
    }
}
