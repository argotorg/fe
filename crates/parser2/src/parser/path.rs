use crate::SyntaxKind;

use super::{define_scope, token_stream::TokenStream, Parser};

define_scope! {
    #[doc(hidden)]
    pub PathScope,
    Path,
    Inheritance
}
impl super::Parse for PathScope {
    fn parse<S: TokenStream>(&mut self, parser: &mut Parser<S>) {
        parser.set_newline_as_trivia(false);
        parser.parse(PathSegmentScope::default(), None);
        while parser.current_kind() == Some(SyntaxKind::Colon2) {
            parser.bump_expected(SyntaxKind::Colon2);
            parser.parse(PathSegmentScope::default(), None);
        }
    }
}

define_scope! {
    PathSegmentScope,
    PathSegment,
    Inheritance
}
impl super::Parse for PathSegmentScope {
    fn parse<S: TokenStream>(&mut self, parser: &mut Parser<S>) {
        match parser.current_kind() {
            Some(kind) if is_path_segment(kind) => {
                parser.bump();
            }
            _ => parser.error_and_recover("expected path segment", None),
        }
    }
}

pub(super) fn is_path_segment(kind: SyntaxKind) -> bool {
    matches!(
        kind,
        SyntaxKind::SelfType | SyntaxKind::SelfKw | SyntaxKind::Ident
    )
}
