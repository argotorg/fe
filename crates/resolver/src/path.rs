use camino::Utf8PathBuf;
use serde::Deserialize;

use crate::remote::GitDescription;

#[derive(Deserialize, Hash, Debug, Clone, PartialEq, Eq)]
pub struct PathDescription {
    pub path: String,
}

#[derive(Clone, Debug)]
enum PathHead {
    Local(Utf8PathBuf),
    Remote(GitDescription),
}

#[derive(Clone, Debug)]
pub struct FullPathDescription {
    head: PathHead,
    sub_paths: Vec<PathDescription>,
}

impl FullPathDescription {
    pub fn new_local(local_path: &Utf8PathBuf) -> Self {
        FullPathDescription {
            head: PathHead::Local(local_path.clone()),
            sub_paths: vec![],
        }
    }

    pub fn new_remote(git_description: &GitDescription) -> Self {
        FullPathDescription {
            head: PathHead::Remote(git_description.clone()),
            sub_paths: vec![],
        }
    }

    pub fn path(&self) -> Utf8PathBuf {
        match &self.head {
            PathHead::Local(local_path) => {
                self.sub_paths
                    .iter()
                    .fold(local_path.clone(), |acc, path_description| {
                        acc.join(Utf8PathBuf::from(path_description.path.to_owned()))
                            .canonicalize_utf8()
                            .unwrap()
                    })
            }
            PathHead::Remote(git_description) => self.sub_paths.iter().fold(
                git_description.relative_path(),
                |acc, path_description| {
                    acc.join(Utf8PathBuf::from(path_description.path.to_owned()))
                        .canonicalize_utf8()
                        .unwrap()
                },
            ),
        }
    }

    pub fn push_sub_path(&self, path_description: &PathDescription) -> Self {
        let mut sub_paths = self.sub_paths.clone();
        sub_paths.push(path_description.clone());
        Self {
            head: self.head.clone(),
            sub_paths,
        }
    }
}
