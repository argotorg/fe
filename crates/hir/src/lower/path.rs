use fe_parser2::ast;

use crate::{
    hir_def::{IdentId, MaybeInvalid, PathId, PathSegment},
    HirDb,
};

impl PathId {
    pub(crate) fn from_ast(db: &dyn HirDb, ast: ast::Path) -> Self {
        let mut segments = Vec::new();
        for seg in ast.into_iter() {
            let segment = if seg.is_self() {
                MaybeInvalid::Valid(PathSegment::Self_)
            } else if seg.is_self_ty() {
                MaybeInvalid::Valid(PathSegment::SelfTy)
            } else if let Some(ident) = seg.ident() {
                MaybeInvalid::Valid(PathSegment::Ident(IdentId::new(
                    db,
                    ident.text().to_string(),
                )))
            } else {
                MaybeInvalid::invalid()
            };
            segments.push(segment);
        }

        Self::new(db, segments)
    }

    pub(crate) fn maybe_from_ast(db: &dyn HirDb, ast: Option<ast::Path>) -> MaybeInvalid<Self> {
        ast.map(|ast| Self::from_ast(db, ast)).into()
    }
}
