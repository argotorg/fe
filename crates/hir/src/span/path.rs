use parser::ast;

use super::define_lazy_span_item;

define_lazy_span_item!(
    LazyPathSpan,
    ast::Path,
    @idx {
        (segment, LazyPathSegmentSpan),
    }
);

define_lazy_span_item!(LazyPathSegmentSpan);
