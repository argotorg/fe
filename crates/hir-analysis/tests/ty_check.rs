mod test_db;
use std::path::Path;

use dir_test::{dir_test, Fixture};
use fe_compiler_test_utils::snap_test;
use fe_hir_analysis::ty::ty_check::check_func_body;
use test_db::HirAnalysisTestDb;

#[dir_test(
    dir: "$CARGO_MANIFEST_DIR/test_files/ty_check",
    glob: "*.fe"
)]
fn test_standalone(fixture: Fixture<&str>) {
    let mut db = HirAnalysisTestDb::default();
    let path = Path::new(fixture.path());
    let file_name = path.file_name().and_then(|file| file.to_str()).unwrap();
    let (top_mod, mut prop_formatter) = db.new_stand_alone(file_name, fixture.content());

    for &func in top_mod.all_funcs(&db) {
        let Some(body) = func.body(&db) else {
            continue;
        };

        let typed_body = check_func_body(&db, func);
        for expr in body.exprs(&db).keys() {
            let ty = typed_body.expr_ty(&db, expr);
            prop_formatter.push_prop(
                func.top_mod(&db),
                expr.lazy_span(body).into(),
                ty.pretty_print(&db).to_string(),
            );
        }
    }

    let res = prop_formatter.finish(&db);
    snap_test!(res, fixture.path());
}
