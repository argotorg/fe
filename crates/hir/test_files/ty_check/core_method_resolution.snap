---
source: crates/hir-analysis/tests/ty_check.rs
expression: res
input_file: test_files/ty_check/core_method_resolution.fe
---
note: 
   ┌─ core_method_resolution.fe:3:17
   │  
 3 │   fn f() -> usize {
   │ ╭─────────────────^
 4 │ │     let x = Option::Some(10)
 5 │ │     let y = Option::default()
 6 │ │ 
   · │
11 │ │     }
12 │ │ }
   │ ╰─^ usize

note: 
  ┌─ core_method_resolution.fe:4:9
  │
4 │     let x = Option::Some(10)
  │         ^ Option<usize>

note: 
  ┌─ core_method_resolution.fe:4:13
  │
4 │     let x = Option::Some(10)
  │             ^^^^^^^^^^^^ fn Some<usize>

note: 
  ┌─ core_method_resolution.fe:4:13
  │
4 │     let x = Option::Some(10)
  │             ^^^^^^^^^^^^^^^^ Option<usize>

note: 
  ┌─ core_method_resolution.fe:4:26
  │
4 │     let x = Option::Some(10)
  │                          ^^ usize

note: 
  ┌─ core_method_resolution.fe:5:9
  │
5 │     let y = Option::default()
  │         ^ Option<usize>

note: 
  ┌─ core_method_resolution.fe:5:13
  │
5 │     let y = Option::default()
  │             ^^^^^^^^^^^^^^^ fn default<Option<usize>>

note: 
  ┌─ core_method_resolution.fe:5:13
  │
5 │     let y = Option::default()
  │             ^^^^^^^^^^^^^^^^^ Option<usize>

note: 
   ┌─ core_method_resolution.fe:7:5
   │  
 7 │ ╭     if y.is_some() {
 8 │ │         y.unwrap()
 9 │ │     } else {
10 │ │         x.unwrap_or_default()
11 │ │     }
   │ ╰─────^ usize

note: 
  ┌─ core_method_resolution.fe:7:8
  │
7 │     if y.is_some() {
  │        ^ Option<usize>

note: 
  ┌─ core_method_resolution.fe:7:8
  │
7 │     if y.is_some() {
  │        ^^^^^^^^^^^ bool

note: 
  ┌─ core_method_resolution.fe:7:20
  │  
7 │       if y.is_some() {
  │ ╭────────────────────^
8 │ │         y.unwrap()
9 │ │     } else {
  │ ╰─────^ usize

note: 
  ┌─ core_method_resolution.fe:8:9
  │
8 │         y.unwrap()
  │         ^ Option<usize>

note: 
  ┌─ core_method_resolution.fe:8:9
  │
8 │         y.unwrap()
  │         ^^^^^^^^^^ usize

note: 
   ┌─ core_method_resolution.fe:9:12
   │  
 9 │       } else {
   │ ╭────────────^
10 │ │         x.unwrap_or_default()
11 │ │     }
   │ ╰─────^ usize

note: 
   ┌─ core_method_resolution.fe:10:9
   │
10 │         x.unwrap_or_default()
   │         ^ Option<usize>

note: 
   ┌─ core_method_resolution.fe:10:9
   │
10 │         x.unwrap_or_default()
   │         ^^^^^^^^^^^^^^^^^^^^^ usize
